File: 2216-delete-the-middle-node-of-a-linked-list/2216-delete-the-middle-node-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    // int getSize(ListNode* head){ // why pointer?
    //     int size = 1;
    //     ListNode* curNode = head;
    //     while(curNode->next != nullptr){
    //         curNode = curNode->next;
    //         size++;
    //     }
    //     return size;
    // }

    // ListNode* deleteMiddle(ListNode* head) {
    //     int size = getSize(head);
    //     if (size == 1){
    //         return nullptr;
    //     } else {
    //         int delId = size/2;
    //         ListNode* curNode = head;
    //         int curId = 0;
    //         while (curId < (delId-1)){
    //             curNode = curNode->next;
    //             curId++;
    //         }
    //         ListNode* delNode = curNode->next;
    //         curNode->next = delNode->next;
    //         delete delNode;
    //         return head;
    //     }
    // }
    ListNode* deleteMiddle(ListNode* head) {
        if(!head->next) return NULL; // If next of head is NULL then we are deleting the first and the only node in LinkedList.
        ListNode* slow=head;    //slow ptr will be at head
        ListNode* fast=head->next;   //fast ptr will be at next of head
        ListNode* n=NULL;
        while(fast && fast->next)
        {
            n=slow;
            slow=slow->next;
            fast=fast->next->next;
        }

        // after the loop there will be two cases as mentioned
        if(fast==NULL) //it means the list is of odd size and slow pointer is at the node to be deleted.
        {
            ListNode* m=slow->next;
            n->next=m;
            delete slow;
            return head;
        }
        if(fast->next==NULL) //this means the list is of even size and slow pointer is exactly at a node before the node to be deleted
        {
            ListNode* m=slow->next;
            slow->next=m->next;
            delete m;
            return head;
        }
        return head;   
    }
};

File: 2216-delete-the-middle-node-of-a-linked-list/2216-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def getSize(self, head) -> int:
        size = 1 # initialize size as 1 (head)
        curNode = head # initialize curNode variable as head
        while curNode.next != None: # if there's next node
            curNode = curNode.next # update curNode to the next node
            size += 1 # and increase size by 1
        return size

    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        size = self.getSize(head)
        if size == 1:
            return None # return sentinel
        else:
            delId = int(size/2) # int() does the rounding-down
            prevNode = None
            curNode = head # initialize curNode as head
            curId = 0 # initialize curIndex as 0
            while curId != delId: # if (while) curId hasn't reached the index to delete
                prevNode = curNode
                curNode = curNode.next # jump to the next node
                curId += 1 # increase curId
            # now curId == delId and curNode is the curId'th node from the head
            # how to remove a node from linked list?
            # set the previous node's next to current node's next 
            # unset current node's next
            prevNode.next = curNode.next
            curNode.next = None
            return head
            
